### DISCUSSION TITLE
Use pydantic model with form data

### DISCUSSION THREAD

**Discussion by ashish01** [fastapi/fastapi#5951] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import Depends, FastAPI, HTTPException, status, Form
from fastapi.security import HTTPBasic, HTTPBasicCredentials
from fastapi.responses import RedirectResponse
import secrets
from fastapi.responses import HTMLResponse
from typing import Dict
from pydantic import BaseModel
from enum import Enum


class UserData(BaseModel):
    name: str = Form(...)
    age: int = Form(...)


user_data = {}


app = FastAPI()


@app.post("/api/save_user")
def api_save_user(data: UserData):
    user_data[data.name] = data.age


@app.post("/save_user")
def save_user(data: UserData = Depends()):
    api_save_user(data)
    return RedirectResponse("/users", status_code=status.HTTP_302_FOUND)


@app.get("/users")
def get_users():
    html_content = """
    <html>
        <head>
            <title>Some HTML in here</title>
        </head>
        <body>
            <h1>Here are all users if any</h1>
            <div>{}</div>
            <a href="/add_user">Add User</a>
        </body>
    </html>
    """.format(
        ", ".join(user_data.keys())
    )
    return HTMLResponse(content=html_content, status_code=200)


@app.get("/add_user")
def add_user():
    html_content = f"""
    <html>
        <head>
            <title>Some HTML in here</title>
        </head>
        <body>
            <form action="/save_user" method="post">
            <div class="form-example">
                <label for="name">Enter your name: </label>
                <input type="text" name="name" id="name" required>
            </div>
            <div class="form-example">
                <label for="age">Enter your age: </label>
                <input type="text" name="age" id="age" required>
            </div>
            <div class="form-example">
                <input type="submit" value="Subscribe!">
            </div>
            </form>
        </body>
    </html>
    """
    return HTMLResponse(content=html_content, status_code=200)
```

### Description

Getting 422 in response instead of 200

### Operating System

macOS

### FastAPI Version

0.89.1

### Python Version

Python 3.11.1

**Comment by fabioueno** [comment]
Hey, @ashish01!

There are a few things that are preventing you from getting what you want here. But one thing to understand is that Pydantic is used for validating data in your models, and FastAPI is used for creating the API. Why am I saying this? `Form` is from `FastAPI`, and is used to [request the form data](https://fastapi.tiangolo.com/tutorial/request-forms). You're mixing things up. 😅

So, with that said, the first thing you'll need to do is fix the `UserData` class - no more `Form` here:

```python
class UserData(BaseModel):
    name: str
    age: int
```

After that, you'll need to change the way you get the parameters from in the endpoint:

```python
@app.post("/save_user")
def save_user(name: str = Form(...), age: int = Form(...)):
    api_save_user(UserData(name=name, age=age))
    return RedirectResponse("/users", status_code=status.HTTP_302_FOUND)
```

Note that I had to instantiate a `UserData` manually since that's what your `api_save_user` function requires.

And that should do the trick. 😉

**Reply by nic-hartley** [reply]
Hi, just making sure I understand -- There **isn't** a way to read a form body into a Pydantic model, then? I have to retype every single form field as a parameter and then call the constructor myself with all those parameters?

**Reply by hasansezertasan** [reply]
> Hi, just making sure I understand -- There **isn't** a way to read a form body into a Pydantic model, then? I have to retype every single form field as a parameter and then call the constructor myself with all those parameters?

That's the problem here.

**Reply by C0oo1D** [reply, marked as answer]
I don’t know how it was before, now it works for me through typing.Annotated:
```
@app.post("/save_user")
def save_user(user_data: Annotated[UserData, Form()]):
    api_save_user(user_data)
    return RedirectResponse("/users", status_code=status.HTTP_302_FOUND)
```

**Comment by ashish01** [comment]
Thanks so much for explaining with such detail and care.

> You're mixing things up.

Indeed, even as I adding Form in my data model it felt like I am breaking abstractions. Why does my data model know how to get the data?

```
@app.post("/save_user")
def save_user(name: str = Form(...), age: int = Form(...)):
    api_save_user(UserData(name=name, age=age))
    return RedirectResponse("/users", status_code=status.HTTP_302_FOUND)
```

What you suggested does work! But in this approach I am not able to leverage Pydantic's validation facilities. Let me change the model a little to explain

```
class UserData(BaseModel):
    name: str
    age: PositiveInt
```

Now we are enforcing the constraint that the age has to be positive. But the form data has not such validation, it will just accept anything. 

```
@app.post("/save_user")
def save_user(name=Form(), age=Form()):
    api_save_user(UserData(name=name, age=age))
    return RedirectResponse("/users", status_code=status.HTTP_302_FOUND)
```


So when I use the age from Form to instantiate the UserData model it will throw a ValidationException. This bubbles up as a 500 internal server error. Of course in case of the API it returns with 422 error code telling that age needs to be positive.

Maybe I can do something like

```
@app.post("/save_user")
def save_user(name=Form(), age=Form()):
    try:
        user_data = UserData(name=name, age=age)
    except ValidationError as e:
        ... convert exception to 422 response ...
    api_save_user(user_data)
    return RedirectResponse("/users", status_code=status.HTTP_302_FOUND)
```

Maybe there is already some utility code in FastApi to convert validation error to 422 response? Thanks!

**Comment by fabioueno** [comment]
No problem at all! I'm happy to help! 😉

About the validation... You may not be able to leverage Pydantic's validation on the endpoint, but FastAPI has its own! [Here](https://fastapi.tiangolo.com/tutorial/query-params-str-validations) you can see some validation regarding string stuff, such as `max_length`. [Here](https://fastapi.tiangolo.com/tutorial/path-params-numeric-validations), on the other hand, there are some numeric ones, such as the `ge` or `gt`, which would allow for checking if a number is positive, for instance. And although they're documented for query and path parameters, they're also defined for [`Form` parameters](https://github.com/tiangolo/fastapi/blob/master/fastapi/params.py#L279-L298)!

One thing I'm thinking it's kinda odd is that you're getting a 500 error... I made the change you proposed, and I'm getting a 422 here... Did you add anything else to your code? 🤔

Oh, and last thing: since you talked about converting errors, you might enjoy the docs section that explains [error handling](https://fastapi.tiangolo.com/tutorial/handling-errors), in particular, [Use the `RequestValidationError` body](https://fastapi.tiangolo.com/tutorial/handling-errors/#use-the-requestvalidationerror-body).

Once again, hope it helps! 🙂

**Reply by fabioueno** [reply]
By the way, sorry, forgot to show how the change would be in the endpoint:
```python
@app.post("/save_user")
def save_user(name: str = Form(), age: int = Form(ge=0)):
    api_save_user(UserData(name=name, age=age))
    return RedirectResponse("/users", status_code=status.HTTP_302_FOUND)
```

This means "greater than or equal to". In case you don't want equality you can change the `ge` to `gt`, and only values greater than zero should be allowed. 😉

**Comment by ashish01** [comment]
>  I made the change you proposed, and I'm getting a 422 

Wow. Here is the full code I am using

```python
from fastapi import Depends, FastAPI, HTTPException, status, Form
from fastapi.security import HTTPBasic, HTTPBasicCredentials
from fastapi.responses import RedirectResponse
import secrets
from fastapi.responses import HTMLResponse
from typing import Dict
from pydantic import BaseModel, PositiveInt
from enum import Enum


class UserData(BaseModel):
    name: str
    age: PositiveInt


user_data = {}


app = FastAPI()


@app.post("/api/save_user")
def api_save_user(data: UserData):
    user_data[data.name] = data.age


@app.post("/save_user")
def save_user(name=Form(), age=Form()):
    api_save_user(UserData(name=name, age=age))
    return RedirectResponse("/users", status_code=status.HTTP_302_FOUND)


@app.get("/users")
def get_users():
    html_content = """
    <html>
        <head>
            <title>Some HTML in here</title>
        </head>
        <body>
            <h1>Here are all users if any</h1>
            <div>{}</div>
            <a href="/add_user">Add User</a>
        </body>
    </html>
    """.format(
        ", ".join(user_data.keys())
    )
    return HTMLResponse(content=html_content, status_code=200)


@app.get("/add_user")
def add_user():
    html_content = f"""
    <html>
        <head>
            <title>Some HTML in here</title>
        </head>
        <body>
            <form action="/save_user" method="post">
            <div class="form-example">
                <label for="name">Enter your name: </label>
                <input type="text" name="name" id="name" required>
            </div>
            <div class="form-example">
                <label for="age">Enter your age: </label>
                <input type="text" name="age" id="age" required>
            </div>
            <div class="form-example">
                <input type="submit" value="Subscribe!">
            </div>
            </form>
        </body>
    </html>
    """
    return HTMLResponse(content=html_content, status_code=200)
```

after running `uvicorn main:app` I first navigate to http://127.0.0.1:8000/add_user and I see

<img width="401" alt="image" src="https://user-images.githubusercontent.com/151985/216233474-2bd914e1-7d1f-4783-99c9-a28ab65ee16b.png">

where I enter some semantically incorrect data

<img width="367" alt="image" src="https://user-images.githubusercontent.com/151985/216233559-e86aaf6b-796d-4ce6-96e8-859799359a61.png">

On submitting I see internal server error

<img width="311" alt="image" src="https://user-images.githubusercontent.com/151985/216233626-3be601dc-cc91-4407-ad70-5b9ccdc5393f.png">

And server logs show this

```
  File "/Users/ashishsharma/.venv/webdev/lib/python3.11/site-packages/fastapi/routing.py", line 237, in app
    raw_response = await run_endpoint_function(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ashishsharma/.venv/webdev/lib/python3.11/site-packages/fastapi/routing.py", line 165, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ashishsharma/.venv/webdev/lib/python3.11/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ashishsharma/.venv/webdev/lib/python3.11/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ashishsharma/.venv/webdev/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
           ^^^^^^^^^^^^
  File "/Users/ashishsharma/.venv/webdev/lib/python3.11/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ashishsharma/Code/fastapi-auth-depends/main.py", line 29, in save_user
    api_save_user(UserData(name=name, age=age))
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "pydantic/main.py", line 342, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 1 validation error for UserData
age
  ensure this value is greater than 0 (type=value_error.number.not_gt; limit_value=0)
```

`def save_user(name: str = Form(), age: int = Form(ge=0)):` works! Thanks again!

**Comment by fabioueno** [comment]
Oh, I see!

There's a subtle difference between your code and my first proposition that may be causing your `500` error. You were using `def save_user(name=Form(), age=Form()):` to define your function, and even without validating that the function should be positive, I already type hinted `age` as an `int` from the start:

```python
def save_user(name: str = Form(), age: int = Form()):
```

See? This way, it's already an `int`.

But I'm glad `def save_user(name: str = Form(), age: int = Form(ge=0))` worked for you!! :partying_face:
